src/main/java/org/opensearch/knn/index/engine/MethodComponent.java

```diff
@@ -328,19 +330,44 @@ public static Map<String, Object> getParameterMapWithDefaultsAdded(
         Map<String, Object> parametersWithDefaultsMap = new HashMap<>();
         Map<String, Object> userProvidedParametersMap = methodComponentContext.getParameters();
         Version indexCreationVersion = knnMethodConfigContext.getVersionCreated();
+        Mode mode = knnMethodConfigContext.getMode();
+        CompressionLevel compressionLevel = knnMethodConfigContext.getCompressionLevel();
+
+        // Check if the mode is ON_DISK and the compression level is one of the binary quantization levels (x32, x16, or x8).
+        // This determines whether to use binary quantization-specific values for parameters like ef_search and ef_construction.
+        boolean isOnDiskWithBinaryQuantization = mode == Mode.ON_DISK
+            && (compressionLevel == CompressionLevel.x32
+                || compressionLevel == CompressionLevel.x16
+                || compressionLevel == CompressionLevel.x8);
+
         for (Parameter<?> parameter : methodComponent.getParameters().values()) {
             if (methodComponentContext.getParameters().containsKey(parameter.getName())) {
                 parametersWithDefaultsMap.put(parameter.getName(), userProvidedParametersMap.get(parameter.getName()));
             } else {
                 // Picking the right values for the parameters whose values are different based on different index
                 // created version.
                 if (parameter.getName().equals(KNNConstants.METHOD_PARAMETER_EF_SEARCH)) {
-                    parametersWithDefaultsMap.put(parameter.getName(), IndexHyperParametersUtil.getHNSWEFSearchValue(indexCreationVersion));
+                    if (isOnDiskWithBinaryQuantization) {
+                        parametersWithDefaultsMap.put(parameter.getName(), IndexHyperParametersUtil.getBinaryQuantizationEFSearchValue());
+                    } else {
+                        parametersWithDefaultsMap.put(
+                            parameter.getName(),
+                            IndexHyperParametersUtil.getHNSWEFSearchValue(indexCreationVersion)
+                        );
+                    }
```

Comment:
Would prefer this logic inside `IndexHyperParametersUtils` for better readability and testability. It can take compression and version (maybe mode if required) and give the right values

Reply:
Make Sense , but I would keep this PR same as current flow , Would take it after this release.